Add driver for MAX42500

kentlibe opened the pull request:
## PR Description

- This adds new driver for the MAX42500 Industrial Power System Monitor Family, along with corresponding device tree bindings.
- Datasheet [MAX42500](https://www.analog.com/media/en/technical-documentation/data-sheets/max42500.pdf)
- Tested in RPI4 with MAX42500_EVKIT_P1.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented:
Is this still WIP? As it stands it's not really ready for reviewing... Please make sure to address CI complains and to re-organize your commits. Merge commits don't really make sense (rebase your dev branch) and please run `git lon --oneline <DIR_YOU'RE_CHANGING>` so you can the git log style for that subsystem. Then, please follow the same style. Also note that a commit as `Cleanup MAX42500 Linux driver.` is not fine. You need to explain what you're actually cleaning and why...

kentlibe commented:
v2
- ready for review
- address CI complaints
- address style issues
- re-organize commits
- rebase dev branch

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+	if (st->config->pece) {
+		/* Compute CRC over entire I2C frame */
+		pece_value = crc8(max42500_crc8, i2c_data,
+							(MAX42500_I2C_RD_FRAME_SIZE - 1), 0);
```
reduce the number of tabs to align this correctly. I think you are using tab with equivalent of 4 spaces in your IDE. Change it to 8-space tab so you can see it correctly.

kentlibe replied:
ok updated tab settings from 4 to 8 spaces

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max42500_reg_write(struct max42500_state *st,
+								u8 reg_addr, u8 data)
```
check your tabs

kentlibe replied:
ok updated tab settings

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max42500_reg_update(struct max42500_state *st,
+								u8 reg_addr, u8 mask, u8 data)
```
check your tabs

kentlibe replied:
ok updated tab settings

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max42500_set_nominal_voltage(struct max42500_state *st,
+										enum max42500_vm_input vm_in,
```
check your tabs

kentlibe replied:
ok updated tab settings

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+ */
+static int max42500_set_nominal_voltage(struct max42500_state *st,
+										enum max42500_vm_input vm_in,
+										u8 voltage)
```
check your tabs

kentlibe replied:
ok updated tab settings

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+	case MAX42500_VM5:
+		st->ov_thresh1[vm_in] = thresh;
+		return max42500_reg_update(st,
+								MAX42500_REG_OVUV1 + vm_in,
+								GENMASK(7, 4),
+								FIELD_PREP(GENMASK(7, 4),
+								thresh));
```
check your tabs

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int max42500_get_comp_status(struct max42500_state *st,
+									u8 vm_in, u8 *status)
```
check your tabs

kentlibe replied:
ok updated tab settings

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+	}
+
+	st->nominal_volt[vm_in] = voltage;
+	return max42500_reg_write(st, reg_addr, voltage);
```
add a new line before this return for better readability.

kentlibe replied:
ok added new line in between

kister-jimenez commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+};
+
+static const struct hwmon_channel_info *max42500_info[] = {
+	HWMON_CHANNEL_INFO(chip,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY |
+			HWMON_C_TEMP_RESET_HISTORY | HWMON_C_POWER_RESET_HISTORY |
+			HWMON_C_REGISTER_TZ | HWMON_C_UPDATE_INTERVAL | HWMON_C_ALARMS,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY,
+			HWMON_C_IN_RESET_HISTORY | HWMON_C_CURR_RESET_HISTORY),
+	HWMON_CHANNEL_INFO(in,
+			HWMON_I_LABEL | HWMON_I_INPUT | HWMON_I_LOWEST | HWMON_I_MIN |
+			HWMON_I_HIGHEST | HWMON_I_MAX,
+			HWMON_I_LABEL | HWMON_I_INPUT | HWMON_I_LOWEST | HWMON_I_MIN |
+			HWMON_I_HIGHEST | HWMON_I_MAX,
+			HWMON_I_LABEL | HWMON_I_INPUT | HWMON_I_LOWEST | HWMON_I_MIN,
+			HWMON_I_LABEL | HWMON_I_INPUT | HWMON_I_LOWEST | HWMON_I_MIN,
+			HWMON_I_LABEL | HWMON_I_INPUT | HWMON_I_LOWEST | HWMON_I_MIN,
+			HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL,
+			HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL,
+			HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL,
+			HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL, HWMON_I_LABEL),
```
can you double check the channels? I think there is a mismatch in the count.

kentlibe replied:
updated this using hwmon ABI for sysfs attributes

kister-jimenez reviewed (changes requested):
First review attempt.

I think you need to do some adjustments in you formatting especially the tabs to make the code more readable. I have commented some of them, but I see a lot needs to be adjusted.

kentlibe commented on Documentation/devicetree/bindings/hwmon/adi,max42500.yaml:
```diff
@@ -0,0 +1,42 @@
+      - adi,max42500
+
+  reg:
+    description: I2C address of slave device.
```
ok will remove this

kentlibe commented on Documentation/devicetree/bindings/hwmon/adi,max42500.yaml:
```diff
@@ -0,0 +1,42 @@
+
+  reg:
+    description: I2C address of slave device.
+    minimum: 0x28
+    maximum: 0x2B
```
ok will remove this as well

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+DECLARE_CRC8_TABLE(max42500_crc8);
+
+#define MAX42500_REG_ID					0x00
+#define MAX42500_REG_CONFIG1			0x01
+#define MAX42500_REG_CONFIG2			0x02
```
ok updated tab settings

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+#define MAX42500_REG_RSTMAP				0x04
+#define MAX42500_REG_STATOV				0x05
+#define MAX42500_REG_STATUV				0x06
+#define MAX42500_REG_STATOFF			0x07
```
ok updated tab settings

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+#define MAX42500_REG_OVUV3				0x13
+#define MAX42500_REG_OVUV4				0x14
+#define MAX42500_REG_OVUV5				0x15
+#define MAX42500_REG_FPSSTAT1			0x16
+#define MAX42500_REG_FPSCFG1			0x17
```
ok updated tab settings

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+#define MAX42500_REG_WDCFG2				0x29
+#define MAX42500_REG_WDKEY				0x2A
+#define MAX42500_REG_WDLOCK				0x2B
+#define MAX42500_REG_RSTCTRL			0x2C
```
ok updated tab settings

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+/** X is set based on the pull configuration of the ADDR pin */
+#define MAX42500_ADDR(x)				(0x28 + (x))
+#define MAX42500_SILICON_ID				(0x30)
+#define MAX42500_I2C_WR_FRAME_SIZE		(4)
+#define MAX42500_I2C_RD_FRAME_SIZE		(5)
```
ok updated tab settings

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1053 @@
+#define MAX42500_I2C_WR_FRAME_SIZE		(4)
+#define MAX42500_I2C_RD_FRAME_SIZE		(5)
+
+/** MAX42500 Nominal voltage computation */
+#define MAX42500_VNOM_MAX_VM1_VM4		0xFF	// 3.6875v
+#define MAX42500_VNOM_MAX_VM5			0xFF	// 5.6v
+#define MAX42500_MIN_VNOM				0x00	// 0.5v
+#define MAX42500_VNOM_STEP_VM1_VM4		0x01	// 0.0125v
+#define MAX42500_VNOM_STEP_VM5			0x01	// 0.02v
+
+/** MAX42500 Undervoltage/Overvoltage maximum and minimum thresholds*/
+#define MAX42500_MAX_THRESH_VM1_VM5		0x0F	// OV = 110.0%, UV = 90.0%
+#define MAX42500_MIN_THRESH_VM1_VM5		0x00	// OV = 102.5%, UV = 97.5%
+#define MAX42500_MAX_THRESH_VM6_V7		0xFF	// 1.775v
+#define MAX42500_MIN_THRESH_VM6_V7		0x00	// 0.5v
```
ok updated this using the bitfield macros FIELD_PREP and FIELD_GET that truncates the upper and lower limits

nunojsa commented on Documentation/devicetree/bindings/hwmon/adi,max42500.yaml:
```diff
@@ -0,0 +1,45 @@
+    maxItems: 1
+
+  gpios:
+    maximum: 2
```
this needs a better name to describe this. Should be a string that you use to request the gpiod

kentlibe replied:
should renaming the gpios to 'en0' and 'en1' ok as indicated in the datasheet?

nunojsa commented on Documentation/devicetree/bindings/hwmon/adi,max42500.yaml:
```diff
@@ -0,0 +1,45 @@
+        reg = <0x28>;
+        gpios = <&gpio 0 0>, <&gpio 1 0>;
+      };
+    };
```
The indentations is typically 4 spaces for examples. Make sure to comply with that

kentlibe replied:
ok I will update this

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+/*
+ * MAX42500 - Industrial Power System Monitor
+ *
+ * Copyright 2024 Analog Devices Inc.
```
2025

kentlibe replied:
ok will update copyright to 2025

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+#include <linux/gpio/driver.h>
+#include <linux/i2c.h>
+#include <linux/of.h>
+#include <linux/of_gpio.h>
```
of_gpio.h is very likely not needed

kentlibe replied:
ok though I saw it on other drivers but will remove this header

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+#include <linux/of_gpio.h>
+#include <linux/io.h>
+#include <linux/mutex.h>
+#include <linux/of_device.h>
```
why the of_device?

kentlibe replied:
I saw this in max6650 and max31827 hwmon implementations

nunojsa replied:
Note that being in another drivers does not make it correct :).

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	.probe		= max42500_probe,
+	.id_table	= max42500_id,
+};
+
```
no need for new line

kentlibe replied:
ok will remove the extra lines

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+		.of_match_table = max42500_of_match,
+	},
+	.probe		= max42500_probe,
+	.id_table	= max42500_id,
```
just use spaces here... `.probe = max42500_probe` and so on

kentlibe replied:
should... `.probe =` be aligned with `.id_table =` for readability?

nunojsa replied:
no, just use one space. It won't improve readability that much and it's harder to maintain (though this struct does not change often). Reasoning is that you define a new member you might need to adjust all alignments just adding noise to the diff

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	st->gc.owner = THIS_MODULE;
+	st->gc.base = -1;
+	st->gc.ngpio = st->num_gpios;
+	st->gc.can_sleep = true;
```
Why do you have the above for gpios? Are you getting  gpios and registering a gpiochip... that looks very fishy? You're also no registering the gpiochip anywhere. It's not how it should be done...

kentlibe replied:
in the previous internal discussion here in adgt, it was suggested to add the gpios in the gpiochip.
how should the gpiochip be implemented with the gpios?

kentlibe replied:
What does MFD mean?

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+static long convert_wd_firstup_to_reg(long value, u8 wclo, u8 wopn)
+{
+	return ((value / (wclo + wopn)) - 1) >> 1;
+}
```
I would drop all this one liner functions... Not seeing any added value

kentlibe replied:
should these functions be moved in the caller functions?
should only the one-liner functions be droppped?

nunojsa replied:
I would do so... Just do the math inline

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	if (IS_ERR(regmap))
+		return PTR_ERR(regmap);
+
+	st->regmap = regmap;
```
Why do you need a custom bus? At a first glance, I'm not seeing the need...

kentlibe replied:
the device's i2c supports both non-PEC and PEC that follows smbus 3.0 format.
Should I remove them in regmap?

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	&sensor_dev_attr_en.dev_attr.attr,
+	NULL
+};
+ATTRIBUTE_GROUPS(max42500);
```
Oh no... The gpio attributes don't belong here for sure. For hwmon, look ate userspace ABI:

https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface

And more importantly, use the new API:

https://elixir.bootlin.com/linux/v6.13.1/source/drivers/hwmon/ltc4282.c#L1568

See how the above is supposed to work and to map your attributes in the common userspace hwmon interface... Bear in mind that you can add some custom attributes but you'll have to clearly justify their need. Some could likely just go to debugfs

kentlibe replied:
How should the custom attributes be handled? Should all of them go to the debugfs?
The MAX42500 device has 3 features:
- Voltage Monitors (VM)
- Flexible Power Sequence Recorder (FPSR)
- Windowed Watchdog (WW)

Some functions in the VM, FPSR and WW have no corresponding attributes in the hwmon ABI and sysfs.

For the gpio attributes, should there be a separate attribute group created?
The two gpio attributes are used to trigger the FPSR timestamp recordings.

The latest MAX42500 datasheet release can be found here:
https://www.analog.com/media/en/technical-documentation/data-sheets/max42500.pdf

nunojsa replied:
> For the gpio attributes, should there be a separate attribute group created?

For this, you do not need any attrgroup

> How should the custom attributes be handled? Should all of them go to the debugfs?

Only the ones that make sense for debugging purposes...

> Flexible Power Sequence Recorder (FPSR)

I did not went into much detail about this but this looks like it could either go to debugFS or you could just do some readings at probe to detect any power up failure during probe. And if so, maybe just fail probing (if the part is not reliable in this state).

> Windowed Watchdog (WW)

This one I'm not sure... It seems some settings could be DT properties while some status could go into debugFS. But IIUC, you need to continuously write some registers in the watchdog so that RESET is not asserted? You might ask for feedback on this one when upstreaming. There's also a drivers/watchdog subsystem and if it fits you might need to turn this into a MFD device (for supporting the watchdog bits)

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	return count;
+}
+
+static ssize_t vmon_store(struct device *dev, struct device_attribute *devattr,
```
be consistent... typoically we have the driver name as a prefix for the function names

kentlibe replied:
ok will add the driver name's prefix

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+			       gpiod_get_value(st->gpios[idx]));
+	}
+
+	return len;
```
You do not control gpios like this... What are the pins on the chip supposed to do?

kentlibe replied:
In the datasheet, they are used to control the logic of the `EN0` and `EN1` pins of the device. Each are controlled either by a RC circuit or SW-controlled from the MCU gpios. The pins are primarily used to control the FPSR (Flexible Power Sequence Recorder) of the device and trigger the Power-On/Off (pin `EN0`) and/or Sleep On/Off (pin `EN1`) timestamp recordings of the voltage rails monitored by the device.

kentlibe replied:
The datasheet can be found here:
https://www.analog.com/media/en/technical-documentation/data-sheets/max42500.pdf

nunojsa replied:
SO, EN0 and EN1 are input pins so you can't really control the pins value, direction, etc... That means, this devices does not look like a gpiochip.

So, ENO looks like a reset pin that you would make sure to be high at probe so that the part is on. Typically, one forces a chip reset during probe to make sure the device is in it's default configuration but for devices like this,m that is typically not advised.

EN1 looks like something you could support as part of the PM subystem. Like this:

https://elixir.bootlin.com/linux/v6.13.2/source/drivers/hwmon/ltc2947-core.c#L1160

kister-jimenez replied:
Host processor can actually control these values. But the way I see is that this is done at the system level/application level already. And yeah, it can be controlled by external circuit as well so I don't think it should be part of this driver. I think the addition of GPIO here are for monitoring the EN0/EN1 pins states. But the EN0 and EN1 state can be read from CONFIG2 register right?

nunojsa replied:
I mean, you can still control them from the driver but as a gpiod consumer (not a chip). In the same way you do, for example, for a [RESET pin](https://elixir.bootlin.com/linux/v6.13.2/source/drivers/iio/adc/ad9467.c#L1204), you could get these pins and try to control the power states.

nunojsa replied:
>  While the EN1 if low will leave only the OFF comparators on input channels that were enabled as is.

Did not fully understood the above...

> The challenge with implementing the EN0 and EN1 pins is that they have multiple functions in the device:

That is why devicetree exists... You need to specify in DT how the pins are intended to be used as that is not going to change at runtime.

> Ahh I see. If we use PM as well, should the EN1/EN0 have IRQs also so that appropriate state will be set when EN pins are handled externally?

Don't think so. If they are to be handled externally, one would just not give those pins in DT so the driver does nothing with them. This makes tem optional as far as the driver is concerned

kentlibe replied:
>> While the EN1 if low will leave only the OFF comparators on input channels that were enabled as is.

>Did not fully understood the above...

The device has 3 comparators: OV, UV and OFF comparators. And below is the exact line from the datasheet:
`With VMPD set to 1 and EN0 low, the device turns off all comparators to reduce quiescent current. With EN1 low, the OFF comparators on input channels that are enabled are left enabled so that the device can continue to monitor active inputs.`

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	unsigned int reg_val;
+	int ret;
+
+	mutex_lock(&st->lock);
```
If you need a lock, use the new cleanup.h RAII API

kentlibe replied:
ok will use the new RAII API in cleanup.h

nunojsa commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1610 @@
+	switch (attr) {
+	case MAX42500_ATTR_CHIP_LABEL: /* ID */
+	case MAX42500_ATTR_CHIP_NAME: /* CID */
+	case MAX42500_ATTR_CHIP_CONFIG_STATUS: /* CFG2 */
```
The above three look like an example of things that could go into debugFS

kentlibe replied:
how should the debugFS be implemented in hwmon?
an implemented debugFS in source file max20730.c located in hwmon/pmbus directory but not sure if this is how it should be done.

nunojsa replied:
please look at other examples... I can give you one more (from me):

https://elixir.bootlin.com/linux/v6.13.2/source/drivers/hwmon/ltc4282.c#L1682

nunojsa reviewed (commented):
First superficial review... I've done some comments but you already need a big refactor on the driver so i'm stopping for now.

You need to use the proper HWMON interface and you need to look at the standard ABI and make sure to use it. Again, bear in mind that custom attrs will need a very good justification with the maintainer. You also need a docs files for the new driver (I don't think I saw one).

Last points about commit messages:

1) You need a better message in the bindings patch (add a small description of the device and drop the .yaml suffix)
2) On the driver git subject, drop the `drivers: ` part

nunojsa commented on drivers/watchdog/max42500_wdt.c:
```diff
@@ -0,0 +1,561 @@
+	ret = max42500_wdog_set_update(st, wdog, attr);
+	if (ret < 0)
+		goto error;
+error:
```
drop the label

nunojsa commented on drivers/watchdog/max42500_wdt.c:
```diff
@@ -0,0 +1,561 @@
+
+	ret = regmap_read(st->regmap, max42500_reg_wdog[attr], &reg_val);
+	if (ret)
+		goto exit;
```
ditto

nunojsa reviewed (commented):
Also make sure to fix the build issues...

You're also missing a docs file in Documentation/hwmon

kentlibe commented on Documentation/devicetree/bindings/hwmon/adi,max42500.yaml:
```diff
@@ -0,0 +1,58 @@
+      mode state of the device. This can be enabled optionally to control the
+      state of comparators in the voltage monitoring and/or trigger timestamp
+      recording in the flexible power sequence recording functions. Check the
+      state transition conditions in datasheet for more information.
```
yes in the datasheet they for power on-off and sleep on-off pins, do you have suggestions?

nunojsa replied:
That's the name in the datasheet but reading what they actual do I'm not sure if powerdown is really powering down the complete chip. Anyways, I do not have a better name and they are named as in the datasheet, so fine by me

kentlibe commented on drivers/mfd/max42500_mfd.c:
```diff
@@ -0,0 +1,74 @@
+static const struct mfd_cell max42500_cells[] = {
+	{
+		.name = "max42500",
+		.of_compatible = "adi,max42500",
```
ok will update this

kentlibe commented on drivers/mfd/max42500_mfd.c:
```diff
@@ -0,0 +1,74 @@
+		.of_compatible = "adi,max42500",
+	}, {
+		.name = "max42500_wdt",
+		.of_compatible = "adi,max42500_wdt",
```
> Also, use MFD_CELL_NAME() and MFD_CELL_OF()
ok will do but can you share reference? also if I move the gpios to MFD then the function definitions as well?

nunojsa replied:
If you dont know this, it's about time :)

https://elixir.bootlin.com/linux/v6.13.4/source

Then, just look for the symbol and you'll see examples

> also if I move the gpios to MFD then the function definitions as well?

Might depend on the maintainer mood. But I already saw properties that affect a child device being defined in the MFD device (in the bindings) even though it's only parsed by the child device. See an example here:

https://elixir.bootlin.com/linux/v6.13.4/source/Documentation/devicetree/bindings/mfd/adi,adp5585.yaml

gpio-reserved-ranges is a gpio property that is only parsed by the gpio child device. But for that to work, you need to do this in the hwmon driver:

https://elixir.bootlin.com/linux/v6.13.4/source/drivers/gpio/gpio-adp5585.c#L187

And then you get the gpios as usual

kentlibe replied:
do we change the dt bind from hwmon to mfd?

nunojsa replied:
Oh, of course.... I did not realized the bindings were still in hwmon. This is an MFD device so it needs to go to MFD. Now, as I was saying in the above comments, either you specify all properties in the top (MFD) device as the example I gave you or you define them locally in each of the child devices (and then you need MFD_CELL_OF()). But for simplicity, maybe do it at the MFD level and see if the maintainers do not complain

nunojsa replied:
> max42500_wdt

I forgot to mention that underscores are not allowed for compatibles. Make it max42500-wdt

kentlibe replied:
for file naming is it ok to use underscores?

kentlibe replied:
do you have preferred commit sequence of different devices, e.g, hwmon, mfd?

nunojsa replied:
also not the preferred way. Also use '-'

nunojsa replied:
1. Should come dt-bindings;
2. Then MFD device;
3. Then Child devices...

kentlibe replied:
ok will do on the next commits

kentlibe commented on drivers/mfd/max42500_mfd.c:
```diff
@@ -0,0 +1,74 @@
+#include <linux/mfd/max42500.h>
+#include <linux/module.h>
+#include <linux/regmap.h>
+#include <linux/platform_device.h>
```
the `mfd/max42500.h` for the device register macro names and other common defintions, the regmap.h for the i2c regmap bus that will be consumed by hwmon and watchdog, the other headers copied from reference drivers.

nunojsa replied:
I meant the <linux/platform_device.h>. Fairly sure it's not needed

kentlibe commented on drivers/mfd/max42500_mfd.c:
```diff
@@ -0,0 +1,74 @@
+    .probe = max42500_mfd_probe,
+    .remove = max42500_mfd_remove,
+};
+
```
ok missed this

kentlibe commented on drivers/mfd/max42500_mfd.c:
```diff
@@ -0,0 +1,74 @@
+    .driver = {
+        .name = "max42500_mfd",
+        .of_match_table = max42500_mfd_of_match,
+        .pm = &max42500_pm_ops,
```
ok will do

kentlibe commented on drivers/hwmon/Kconfig:
```diff
@@ -1131,6 +1131,19 @@ config MAX31827
 
+config SENSORS_MAX42500
+	tristate "MAX42500 Industrial Power System Monitor Family"
+	depends on I2C
```
so should this then be `depends on MFD_MAX42500`?

nunojsa replied:
Exactly. Then it will inherently depend on I2C

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1300 @@
+					MAX42500_REG_DTIME4,
+					MAX42500_REG_DTIME5,
+					MAX42500_REG_DTIME6,
+					MAX42500_REG_DTIME7 };
```
ok will do

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1300 @@
+	debugfs_create_file_unsafe("chip_bist_rst_enable_log", 0600, dentry, st,
+				   &max42500_chip_bist_rst_enable_log);
+	debugfs_create_file_unsafe("chip_pec_enable_log", 0600, dentry, st,
+				   &max42500_chip_pec_enable_log);
```
thanks for the ref

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1300 @@
+
+	regmap = dev_get_drvdata(pdev->dev.parent);
+	if (IS_ERR(regmap))
+		return PTR_ERR(regmap);
```
ok will remove the checks

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1300 @@
+				   &max42500_in_status_ov_log);
+	debugfs_create_file_unsafe("in_status_uv_log", 0400, dentry, st,
+				   &max42500_in_status_uv_log);
+	debugfs_create_file_unsafe("in_status_off_log", 0400, dentry, st,
```
these are status for the ov, uv and off comparators

kentlibe replied:
hwmon interfaces?

nunojsa replied:
yes, ov and uv could be likely use the critical alarms and off is the enable interface. Again, assuming this no event log and real status of teh device

kentlibe commented on drivers/hwmon/max42500.c:
```diff
@@ -0,0 +1,1300 @@
+static int max42500_store_pwrup_capt_finish_log(void *arg, u64 *val)
+{
+	return max42500_store_fpsr_log(arg, val,
+				      MAX42500_FPSR_POWERUP_SEQ_CAPTURE_FINISH);
```
ok makes sense and will update

kentlibe commented on include/linux/mfd/max42500.h:
```diff
@@ -0,0 +1,61 @@
+#define __MFD_MAX42500_H
+
+#include <linux/mfd/core.h>
+#include <linux/regmap.h>
```
ok will remove them

kentlibe commented:
> Also make sure to fix the build issues...
> 
> You're also missing a docs file in Documentation/hwmon

since we moved the dt-bindings from hwmon to mfd, this follows as well in Documentation/mfd?

nunojsa commented on drivers/watchdog/Kconfig:
```diff
@@ -703,6 +703,17 @@ config MAX77620_WATCHDOG
+config MAX42500_WDT
+	tristate "MAX42500 Windowed Watchdog"
+	depends on MFD_MAX42500 || COMPILE_TEST
+	depends on I2C
```
I would remove it

nunojsa commented:
@kentlibe, please ping me Monday for reviewing your latest changes but I will already leave you with something that needs change. The MFD device is the main device, so the naming should be:

* max42500.c -> for the MFD device;
* max42500-hwmon.c -> HWMON;
* max42500-wtd.c -> Watchdog

Same way of thinking for driver names and compatibles (if applicable)

kentlibe commented:
v4
- split into 3 drivers: mfd, hwmon and watchdog
- implement the mfd as the main device with the i2c driver
- rework the hwmon driver to follow latest ABI for sysfs platform
- rework the watchdog driver to follow watchdog sysfs platform
- add debugfs to handle all non-standard attr in hwmon and watchdog
- implement the two gpios control in power management format
- renamed the 3 driver files according to their usage and purpose.

@nunojsa kindly review.

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+	if (st->fpsr[MAX42500_FPSR_FPS_EN1_START_TIMER_ENABLE])
+		gpiod_set_value(st->sleep_gpio, GPIOD_OUT_HIGH);
+	else
+		gpiod_set_value(st->sleep_gpio, GPIOD_OUT_LOW);
```
Why not just something like?

`gpiod_set_value(st->sleep_gpio, !!st->fpsr[MAX42500_FPSR_FPS_EN1_START_TIMER_ENABLE])`

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+	if (st->fpsr[MAX42500_FPSR_FPS_EN1_START_TIMER_ENABLE])
+		gpiod_set_value(st->sleep_gpio, GPIOD_OUT_LOW);
+	else
+		gpiod_set_value(st->sleep_gpio, GPIOD_OUT_HIGH);
```
ditto

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+	if (IS_ERR(st->sleep_gpio))
+		return PTR_ERR(st->sleep_gpio);
+
+	devm_mutex_init(dev, &st->lock);
```
error handling

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+
+	ret = regmap_write(st->regmap, max42500_reg_fpsr[index], reg_val);
+	if (ret)
+		return ret;
```
This looks a read modify write... Have you checked regmap_update_bits()?

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+		return ret;
+
+	/* Update after successful register write */
+	st->vmon[index] = FIELD_GET(GENMASK(7, 0), value);
```
same comment

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+
+	ret = regmap_write(st->regmap, max42500_reg_chip[index], reg_val);
+	if (ret)
+		return ret;
```
also looks like a RMW

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1339 @@
+					MAX42500_REG_DTIME5,
+					MAX42500_REG_DTIME6,
+					MAX42500_REG_DTIME7
+					};
```
Put this on the beginning of the line

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,574 @@
+	st->client = client;
+	st->regmap = regmap;
+
+	devm_mutex_init(dev, &st->lock);
```
error handling...

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,574 @@
+{
+	struct device *parent_dev = pdev->dev.parent;
+	struct regmap *regmap = dev_get_drvdata(parent_dev);
+	struct i2c_client *client = to_i2c_client(parent_dev);
```
do you need the i2c_client?

kentlibe replied:
will be removed

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,574 @@
+	const u8 clk_div = MAX42500_WDOG_WD_CLOCK_DIVIDER;
+	const u8 win_mode = MAX42500_WDOG_WD_SIMPLE_MODE_ENABLE;
+	const u8 opn_win = MAX42500_WDOG_WD_OPEN_WINDOW_INTERVAL;
+	const u8 clo_win = MAX42500_WDOG_WD_CLOSE_WINDOW_INTERVAL;
```
I would drop the const qualifier. It does not give you much in here and the compiler should be more than capable of realizing it anyways without the qualifier

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,574 @@
+
+	ret = regmap_write(st->regmap, max42500_wdt_reg_wdog[index], reg_val);
+	if (ret)
+		return ret;
```
same comments as for the hwmon driver... seems like regmap_update_bits() could be used

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,574 @@
+
+	ret = regmap_write(st->regmap, MAX42500_REG_WDKEY, reg_val);
+	if (ret)
+		return ret;
```
ditto

kentlibe replied:
For simple mode, yes can update this. But for challenge-response mode, we need to read the current key first, apply the LFSR algorithm and write back the new key. Maybe we could leave this here or go with the change?

nunojsa replied:
Hmm, now that I look better into this, simple mode looks odd. We read the key and then write the same key back. Is this expected?

kentlibe replied:
yes, I see that would be simpler and optimal. So, code would now look like:
```c
	guard(mutex)(&st->lock);

	switch (mode) {
	case MAX42500_WD_MODE_CH_RESP:
		ret = regmap_read(st->regmap, MAX42500_REG_WDKEY, &key);
		if (ret)
			return ret;
		
		/* Linear-Feedback Shift Register (LFSR) Polynomial Equation */
		lfsr = ((key >> 7) ^ (key >> 5) ^ (key >> 4) ^ (key >> 3)) & 1;
		reg_val = (key << 1) | lfsr;

		return regmap_write(st->regmap, MAX42500_REG_WDKEY, reg_val);
	case MAX42500_WD_MODE_SIMPLE:
		return regmap_write(st->regmap, MAX42500_REG_WDKEY, 0);
	default:
		return -EINVAL;
	}

nunojsa reviewed (commented):
Looks better to me... If you want we can still do another internal review. If not, take my latest inputs and send it upstream. Up to you :)

kentlibe committed:
dt-bindings: mfd: add adi,max42500

The MAX42500 component structure of the nodes and properties of the serial
interfaces and the I/O ports to access, control and configure the voltage
monitors, and to program the power sequence timestamp recordings and the
windowed watchdog timing.

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe committed:
mfd: add driver for max42500

The MAX42500 is a multiple function device for power system monitoring,
power sequence timestamp recordings and windowed watchdog combined all-
together in one SoC package. These features can be used independently or
at the same time in a wide range of applications.

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe committed:
doc: add supported sysfs for max42500

The MAX42500 supports standard power management, hwmon and watchdog ABI
and sysfs attribute entries to expose the sensor data of the voltage
monitors, power sequence timestamp recordings and windowed watchdog to
the userspace.

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe committed:
hwmon: add driver for max42500-hwmon

The MAX42500 is a SoC power-system monitor with up to seven voltage
monitor inputs that has programmable OV/UV thresholds and support DVS
by I2C interface. Also, it has a programmable flexible power sequence
recorder that stores the power-up and power-down timestamps separately.

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe committed:
watchdog: add driver for max42500-wdt

The MAX42500 contains a programmable windowed watchdog and can be used
either as simple mode windowed watchdog or challenge-and-response mode
watchdog, which is accessible through the I2C interface, along with a
configurable RESET output.

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe committed:
Kconfig.adi: imply SENSORS_MAX42500

Add entry for the MAX42500 driver

Signed-off-by: Kent Libetario <Kent.Libetario@analog.com>

kentlibe commented on Kconfig.adi:
```diff
@@ -95,6 +95,8 @@ config KERNEL_ALL_ADI_DRIVERS
 	imply REGULATOR_MAX77541
 	imply MFD_MAX77541
+	imply MAX42500_WDT
+	imply MFD_MAX42500
```
this should be together with the hwmon/Kconfig.adi commit

kentlibe commented:
@nunojsa could you take one final review before we upstream?

kentlibe commented:
@kister-jimenez is the dt-binding check error caused by a CI script error?

nunojsa commented:
> @kister-jimenez is the dt-binding check error caused by a CI script error?

Yes, our CI needs to be fixed for the bindings check

nunojsa commented on Documentation/devicetree/bindings/mfd/adi,max42500.yaml:
```diff
@@ -0,0 +1,58 @@
+properties:
+  compatible:
+    enum:
+      - adi,max42500-mfd
```
Typically, you do not need the -mfd for the real device (meaning for the top device). Keep it only for -hwmon and -wdt

nunojsa commented on Documentation/devicetree/bindings/mfd/adi,max42500.yaml:
```diff
@@ -0,0 +1,58 @@
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        hwmon@28 {
```
you'll likely need a more generic name in here since this is not only an hwmon device

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1346 @@
+#include <linux/gpio/consumer.h>
+#include <linux/pm.h>
+#include <linux/i2c.h>
+#include <linux/module.h>
```
I think you need mfd_devicetable.h for `struct platform_device_id`

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1346 @@
+	gpiod_set_value(st->power_gpio, GPIOD_OUT_LOW);
+
+	gpiod_set_value(st->sleep_gpio,
+			st->fpsr[MAX42500_FPSR_FPS_EN1_START_TIMER_ENABLE]);
```
nit: I would do `!!st->fpsr[MAX42500_FPSR_FPS_EN1_START_TIMER_ENABLE]` just to be sure to pass 0 or 1 into gpiod_set_value()

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,550 @@
+#include <linux/device.h>
+#include <linux/regmap.h>
+#include <linux/watchdog.h>
+#include <linux/module.h>
```
same comment for mod_devicetable.h

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,550 @@
+	.driver = {
+		.name = "max42500-wdt",
+	},
+	.probe = max42500_wdt_probe,
```
I would add an .id table as you have in the hwmon driver

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,550 @@
+
+static int max42500_wdt_store_wd_rst2_cnt_enable_log(void *arg, u64 val)
+{
+	return max42500_wdt_wdog_write(arg, (long)val,
```
this explicit cast should not be needed

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,550 @@
+
+static int max42500_wdt_show_wd_rst2_cnt_enable_log(void *arg, u64 *val)
+{
+	return max42500_wdt_wdog_read(arg, (long *)val,
```
not sure this is correct. This will go wrong in big endian 32bit archs... I would use a dedicated variable. This:

```
long __val;
int ret;

ret = max42500_wdt_wdog_read(arg, &__val, ...);
if (ret)
   return ret;

*val = __val; 
```

Same for the other places where this applies. Note this is mostly a problem fro getters and not setters since getters use pointers which this cast to potentially smaller types could be problematic.

kentlibe commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,550 @@
+		return ret;
+
+	/* Update after successful register write */
+	st->wdog[MAX42500_WDOG_WD_KEY] = FIELD_GET(GENMASK(7, 0), reg_val);
```
And I think I would also remove this line.

nunojsa replied:
yeah, no need to cache all values you write even more when you read them from the device

nunojsa commented on Documentation/devicetree/bindings/mfd/adi,max42500.yaml:
```diff
@@ -0,0 +1,58 @@
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        max42500@28 {
```
I do not think this will be accepted... Needs to be more generic but honestly i'm not sure what can we put in here. I know I was the one to mention this but maybe just send this upstream as you had it before. Meaning, "hwmon" since the primary function of the device is monitoring.

kentlibe replied:
ok will do. Previously I wrote "mfd" then "mfdev", but I thought it was too generic since all other mfd drivers are also too.

kentlibe replied:
Saw something similar in richtek and/or ricoh mfd dt-bindings and was not sure if same for max42500. But I guess not.

nunojsa commented on drivers/mfd/max42500.c:
```diff
@@ -0,0 +1,60 @@
+				   max42500_subdev, ARRAY_SIZE(max42500_subdev),
+				   NULL, 0, NULL);
+
+	return ret;
```
you can just `return devm_mfd_add_devices(...)`

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1347 @@
+#include <linux/mutex.h>
+#include <linux/mfd/max42500.h>
+#include <linux/platform_device.h>
+#include <linux/mod_devicetable.h>
```
alphabetical order

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1347 @@
+static struct platform_driver max42500_hwmon_driver = {
+	.driver = {
+		.name	= "max42500-hwmon",
+		.pm = &max42500_pm_ops,
```
use pm_sleep_ptr()

nunojsa commented on drivers/hwmon/max42500-hwmon.c:
```diff
@@ -0,0 +1,1347 @@
+	return 0;
+}
+
+EXPORT_SIMPLE_DEV_PM_OPS(max42500_pm_ops, max42500_suspend, max42500_resume);
```
why exporting this? Do you need this function anywhere else? I guess not...

kentlibe replied:
will remove this

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,596 @@
+	int ret;
+
+	ret = max42500_wdt_wdog_read(watchdog_get_drvdata(wdd), &wdog_val,
+				      MAX42500_WDOG_WD_STATUS);
```
alignment seems a bit off here

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,596 @@
+	guard(mutex)(&st->lock);
+
+	switch (mode) {
+	case MAX42500_WD_MODE_CH_RESP:
```
nit: the mutex is only needed in this case

kentlibe replied:
when the mutex is moved in the case, there is build error:
```bash
In file included from drivers/watchdog/max42500-wdt.c:8:
drivers/watchdog/max42500-wdt.c: In function ‘max42500_wdt_set_watchdog_key’:
./include/linux/cleanup.h:86:2: error: a label can only be part of a statement and a declaration is not a statement
   86 |  class_##_name##_t var __cleanup(class_##_name##_destructor) = \
      |  ^~~~~~
./include/linux/cleanup.h:131:2: note: in expansion of macro ‘CLASS’
  131 |  CLASS(_name, __UNIQUE_ID(guard))
      |  ^~~~~
drivers/watchdog/max42500-wdt.c:94:3: note: in expansion of macro ‘guard’
   94 |   guard(mutex)(&st->lock);
      |   ^~~~~
make[4]: *** [scripts/Makefile.build:243: drivers/watchdog/max42500-wdt.o] Error 1
make[3]: *** [scripts/Makefile.build:480: drivers/watchdog] Error 2
make[3]: *** Waiting for unfinished jobs....
```

nunojsa replied:
You need the scoped version...

nunojsa commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,596 @@
+		reg_val = (key << 1) | lfsr;
+
+		return regmap_write(st->regmap, MAX42500_REG_WDKEY, reg_val);
+	case MAX42500_WD_MODE_SIMPLE:
```
nit: I would be tempted to put a comment stating that we just need to do a random write in this case (just to avoid potential questions) - not sure if the intent is already obvious.

kentlibe replied:
should I add the mutex in the case here too?

nunojsa replied:
no, regmap already has an internal lock unless you disable it in the config (which you're not doing)

kentlibe commented on drivers/watchdog/max42500-wdt.c:
```diff
@@ -0,0 +1,596 @@
+	ret = max42500_wdt_wdog_read(watchdog_get_drvdata(wdd), &wdog_val,
+				      MAX42500_WDOG_WD_STATUS);
+	if (ret)
+		return MAX42500_WDOG_WDSTAT_MASK;
```
added error handling in case regmap fails

nunojsa reviewed (commented):
Feel free to send this upstream. As another advice, I would improve the commit messages for the bindings patch. Saying "adding bindings file" or something like that is only stating the obvious, You should add a short description of the device being added (similar to the driver patch)
